framework/source/class/qx/core/MObjectId.js

```diff
@@ -157,59 +166,85 @@ qx.Mixin.define("qx.core.MObjectId", {
         if (found) {
           result = target;
         }
-        
+
       } else {
         // No object, creating the object
         result = this._createQxObject(id);
       }
-      
+
       if (result && controlId) {
         var childControl = result.getChildControl(controlId);
         return childControl;
       }
-      
+
       return result;
     },
-    
+
     /**
      * Creates the object and adds it to a list; most classes are expected to
-     * override `_createQxObjectImpl` NOT this method.
-     * 
+     * override `_createQxObjectImpl` NOT this method. If `_createQxObjectImpl`
+     * returns `undefined`, this is almost always the result of a misconfiguration
+     * (such as a typo in the id), since `null` is used if you explicitly do
+     * not want to return a value from `_createQxObjectImpl`. To address
+     * this, a warning is emitted if the environment variable `"qx.debug"` is
+     * `true`. In order to make the source of the problem easier to locate,
+     * set the environment variable `"module.objectId.debugger”`
+     * to `true` to trigger the browser debugger.
+     *
      * @param id {String} ID of the object
+     * @param owner {qx.core.Object?} If given, the owner object to which the
+     * created object is attached. Otherwise, the current instance (`this`) is used.
      * @return {qx.core.Object?} the created object
      */
-    _createQxObject: function(id) {
-      var result = this._createQxObjectImpl(id);
+    _createQxObject: function(id, owner) {
+      if (owner === undefined) {
+        owner = this;
+      } else if (!(owner instanceof qx.core.Object)) {
+        throw new Error("Owner must be instance of qx.core.Object or undefined.");
+      }
+      var result;
+      if (typeof this.createQxObject == "function") {
```

Comment:
IMHO this is not a good way to change the API method name because it's non-obvious - the reader would normally expect the list of methods in `members` to define the API, and it is documented that the `_createQxObjectImpl` will be called, yet this change means that if the class *or any of it's descendant classes* implement `createQxObject` then the documented method is ignored.

I do think that in retrospect it would be good to change the naming here, because as @cboulanger says the object ID mechanism supports a coding pattern which is really useful in creating clear and concise widget UIs - more on that later, we've talked about a medium.com blog posting as well as docs to explain our thoughts.

IMHO the "create" method should stay protected though, because part of the API specification is that "create" will only be called once for each ID (and it could easily cause issues if that was not followed).  By making it protected and using the public `getQxObject` to control access, the `getQxObject` is a gatekeeper which reliably enforces those API rules.

I think I would prefer that `_createQxObjectImpl` and `_createQxObject` are simply swapped around, i.e. that the method which people override is called `_createQxObject` and the internal method which `getQxObject` defers to is `_createQxObjectImpl`.

This would be a breaking API change, but as we're talking about pre-release there is room to decide to do this; IMHO the benefits are better clarity because the design pattern that it promotes (or: that we intend to promote) is really very effective.

Reply:
In line with @oetiker's comment below, we could detect the implementation of `_createQxObjectImpl` in debug mode and give a warning - this detection/warning would be scheduled to be removed in 7.x